### ISSUE TITLE
Lost wake-ups due to race

### ISSUE THREAD

**Issue by sbarral** [fereidani/kanal#15]
This is a second issue which is related but not the same as #14. Unlike #14, it is a regular race and not a data race, so AFAICT it does not expose any UB, but it will lead to lost wake-ups (a receiver could never wake up even though data is available). As I write below, I suspect that a variation of this issue exists as well in the `SyncSignal`, but I haven't reviewed the latter.

The problem can be seen for instance in the interaction between these 2 methods:

https://github.com/fereidani/kanal/blob/79162a6f50984229890824706d76d40556cf1511/src/signal.rs#L22-L39

https://github.com/fereidani/kanal/blob/79162a6f50984229890824706d76d40556cf1511/src/signal.rs#L83-L90

A race would appear in this scenario:

1) `poll` reads `LOCKED` at line 26
2) `send` reads an old (no longer actual) waker at line 87
3) poll sets a new waker at line 31 and reads the state again at line 33, but since the read is just a `SeqCst` read, with respect to the waker store (a relaxed operation) it just acts as an `Acquire` operation, meaning it may be reordered to
  a) first: read `LOCKED` again
  b) then: set new waker
4) send stores `UNLOCK` and wakes the old waker: the receive task will never wake up.

As suggested in issue #14, I would strongly urge you to instrument all code with Loom (use `loom::cell::UnsafeCell` etc) as this would definitely have caught these types of races.

On the bright side, any of the 2 solutions I suggested in #14 should also solve this race.

I suspect that a similar issue exists in `SyncSignal`, but this is only based on a gut feeling because `evencount`-type primitives (which were originally invented for classical blocking code such as `SyncSignal`) will nearly always need to use full fences or Read-Modify-Atomic operations to avoid such issues, and I don't see any in your code.

**Comment by fereidani** [comment]
Hi, Thanks for the issue, I found some dead code in the project because of it, But it's incorrect, Waker is registered here before pushing the signal to the signal list:
https://github.com/fereidani/kanal/blob/ac93742f47ce7ac6eb88969274e4e47f05f526a7/src/future.rs#L92-L97
and
https://github.com/fereidani/kanal/blob/ac93742f47ce7ac6eb88969274e4e47f05f526a7/src/future.rs#L179-L185

and this code was a dead code that I forgot to remove:
```rust
     { 
         self.waker = Some(cx.waker().clone()) 
     } 
```
so before pushing the signal to the queue, the waker is already registered, and the possibility of lost wake-ups does not exist.

Feel free to close the issue if you think my explanation is clear.

**Comment by sbarral** [comment]
So as I wrote in the sibling issue: you *need* to re-register the waker every time because it can change every time `poll` is called, so the dead code was actually something that should be done, albeit safely.

**Comment by fereidani** [comment]
Fixed as discussed in #14
